.github/workflows/ci.yml

```diff
@@ -175,3 +175,24 @@ jobs:
     - name: Build eggs
       run: python setup.py bdist_egg
     - run: echo "🥑 This job's status is ${{ job.status }}."
+
+  experimental-checks:
+    name: Experimental checks
+    runs-on: ubuntu-20.04
+    steps:
+      - run: echo "Job triggered by a ${{ github.event_name }} event on branch is ${{ github.ref }} in repository is ${{ github.repository }}, runner on ${{ runner.os }}"
+      - name: Check out repository code
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Codespell Check
+        uses: codespell-project/actions-codespell@master
+        with:
+          check_filenames: true
+          check_hidden: true
+          only_warn: 1
+      - name: Run bandit
+        uses: tj-actions/bandit@v5
+        with:
+          options: '-r --exit-zero'
+      - run: echo "🥑 This job's status is ${{ job.status }}."
```

Comment:
Hi @ana afaict this is welcome, but I'm concerned here how to reproduce this locally. Since they are using public actions, what are the exact commands to run this locally?

Reply:
Install banding with `dnf install bandit` and then: ` bandit . -r --exit-zero`
The action is just much providing a update bandit and letting you to add nicely the options.
More info at the actions page https://github.com/tj-actions/bandit and https://bandit.readthedocs.io/en/latest/plugins/index.html

Reply:
So bandit and other dependencies should be part of `requirements-dev.txt` when this get accepted?

I understand the nice usage of those public actions, but for reproducibility reasons,  what do you think about having targets here that are pointing to Makefile or setup.py targets so would be easier for us during local runs? Example:

```bash
$ python3 setup.py test --security-checks
```

Following the same logic as we have today for "static-checks" and others.

Reply:
We already have some jobs running tests that are hard to reproduce locally, like the Mac smoke test or some Cirrus test. I would not worry much about reproducing it locally.

Reply:
No, no new dependencies!  While we can run both codespell and bandit locally, this is a complementary check to run only in github. We're already doing this with LGTM or packit.
Codespell is extra for pylint's spellchecker that does 99% of the job. Bandit is an alternative and complement to LGTM.com that's good to have because LGTM.com is an external website that we know can disappear any day.

Reply:
We could take this example to better think about it, using it to improve our processes. I'm trying to forget how we work today. I'm just thinking of how we'd like it to be based on previous reproducibility issues.

Even knowing that this is an experiment, I would like to stress how we are going to integrate this soon when it became part of our gating CI.

In the same way that a website can go down and have issues, a public GH-Action can break, cease to be maintained and have security issues as well, especially here, since are pinning the version. We're in both cases depending on third parties, and I'm not against that.

But, sometimes we are having constant issues with third-party providers, and in the rush, we might wish/tend to ignore them, reducing our overall CI credibility.  Because of that, I would like to stress what is necessary to have and what is not.

Until now we were only using actions for "clone", "upload artifact" and "setup-python". None of these actions needed to be performed locally during the patch submission process, which is why I would like to stress this and make sure we are on the right path.

Regarding reproducibility, IMO is an awful experience for the developer to run local tests, and have some tests captured only in some further stages of CI. 

(This weekend, we had an example of this before the release and we had to solve last-minute problems.)

I do understand that we won't be able to reproduce all stages of our CI locally, because of architectures or some specific tests we chose to run in a further step. However, some tests such as grammatical checks do not need to wait for such rounds and the sooner we detect the better.

Nevertheless, as long as they are part of the experimental and non-mandatory tests for PR to enter, I'm ok with that. I just wanted to bring up the discussion, as I imagine that the fewer complementary tests we have, the better will be since we should try to eliminate as much as possible the "optional tests". Once there, we tend to forget about them (since they are not gating). I also understood that the purpose of these tests would be to be part of the gating one day, and I would like to discuss it.

Reply:
@beraldoleal We have discussed this during a meeting, so I'll be brief.

These test won't be mandatory for the PR to be approved, they give some extra information that can be used to improve the PR.
I'm not expecting contributors to run this locally before submitting the PR, however they'll have the pipeline available in their fork and should be able to run it before submitting.
I can rename it from "experimental checks" to "extra checks" if that make things easier.
I'll add a quick comment over every test explaining how to run it locally (if possible), it's fast to do and you never know when it can be helpful.

I was also concerned about reproductibility, sooner or later, GitHub will restrict the usage for free and we'll have the same problem we got with Travis.
I have filed an issue to remember me to look at this https://github.com/avocado-framework/avocado/issues/4926

Reply:
> @beraldoleal We have discussed this during a meeting, so I'll be brief.
> 
> These test won't be mandatory for the PR to be approved, they give some extra information that can be used to improve the PR.
> I'm not expecting contributors to run this locally before submitting the PR, however they'll have the pipeline available in their fork and should be able to run it before submitting.
> I can rename it from "experimental checks" to "extra checks" if that make things easier.

I had understood that those checks would at some point be part of our pipeline as gating, and that was the reason I brought this up. If they are being flagged as "extra" now and we have no intentions to make this "part of our pipeline", as I said before, I'm ok with this.

However, TBH I was expecting having this as part of our pipeline, since the tools are better and are catching more errors (at least the spell ones). So was happy with the new tools and hopping to promoting them to the PR's checks soon. I was just concerned about the "reproducibility" aspect.

> I'll add a quick comment over every test explaining how to run it locally (if possible), it's fast to do and you never know when it can be helpful.
> 
> I was also concerned about reproductibility, sooner or later, GitHub will restrict the usage for free and we'll have the same problem we got with Travis.
> I have filed an issue to remember me to look at this #4926

That is good you are thinking about this. Thanks. I will take a look it.

Reply:
@beraldoleal codespell is not better than pylint/enchant, it just work differently. It checks for common misspellings in text  and do not have false positives while pylint is more exhaustive at the cost of not catching some misspells.

bandit, however, provides some security checking we don't have right now and depending how well it works, i might propose to "promote" it to mandatory check.

Reply:
FIY, and FWIW, there are bandit based checks in code climate.